internal/resolution/entitysources/catalogdsource.go

```diff
@@ -71,70 +72,117 @@ func (es *CatalogdEntitySource) Iterate(ctx context.Context, fn input.IteratorFu
 	return nil
 }
 
-func getEntities(ctx context.Context, client client.Client) (input.EntityList, error) {
-	entityList := input.EntityList{}
-	bundleMetadatas, packageMetdatas, err := fetchMetadata(ctx, client)
-	if err != nil {
+func getEntities(ctx context.Context, cl client.Client) (input.EntityList, error) {
+	allEntitiesList := input.EntityList{}
+
+	var catalogList catalogd.CatalogList
+	if err := cl.List(ctx, &catalogList); err != nil {
 		return nil, err
 	}
-	for _, bundle := range bundleMetadatas.Items {
-		props := map[string]string{}
-
-		// TODO: We should make sure all properties are forwarded
-		// through and avoid a lossy translation from FBC --> entity
-		for _, prop := range bundle.Spec.Properties {
-			switch prop.Type {
-			case property.TypePackage:
-				// this is already a json marshalled object, so it doesn't need to be marshalled
-				// like the other ones
-				props[property.TypePackage] = string(prop.Value)
-			case entities.PropertyBundleMediaType:
-				props[entities.PropertyBundleMediaType] = string(prop.Value)
-			}
+	for _, catalog := range catalogList.Items {
+		channels, bundles, err := fetchCatalogMetadata(ctx, cl, catalog.Name)
+		if err != nil {
+			return nil, err
 		}
 
-		imgValue, err := json.Marshal(bundle.Spec.Image)
+		catalogEntitiesList, err := MetadataToEntities(catalog.Name, channels, bundles)
 		if err != nil {
 			return nil, err
 		}
-		props[entities.PropertyBundlePath] = string(imgValue)
-		catalogScopedPkgName := fmt.Sprintf("%s-%s", bundle.Spec.Catalog.Name, bundle.Spec.Package)
-		bundlePkg := packageMetdatas[catalogScopedPkgName]
-		for _, ch := range bundlePkg.Spec.Channels {
-			for _, b := range ch.Entries {
-				catalogScopedEntryName := fmt.Sprintf("%s-%s", bundle.Spec.Catalog.Name, b.Name)
-				if catalogScopedEntryName == bundle.Name {
-					channelValue, _ := json.Marshal(property.Channel{ChannelName: ch.Name, Priority: 0})
-					props[property.TypeChannel] = string(channelValue)
-					replacesValue, _ := json.Marshal(entities.ChannelEntry{
-						Name:     b.Name,
-						Replaces: b.Replaces,
-					})
-					props[entities.PropertyBundleChannelEntry] = string(replacesValue)
-					entity := input.Entity{
-						ID:         deppy.IdentifierFromString(fmt.Sprintf("%s%s%s", bundle.Name, bundle.Spec.Package, ch.Name)),
-						Properties: props,
-					}
-					entityList = append(entityList, entity)
+
+		allEntitiesList = append(allEntitiesList, catalogEntitiesList...)
+	}
+
+	return allEntitiesList, nil
+}
+
+func MetadataToEntities(catalogName string, channels []declcfg.Channel, bundles []declcfg.Bundle) (input.EntityList, error) {
```

Comment:
Could you please explain your concerns?

I'm concerned about `CatalogdEntitySource` since it seems to call `getEntities` on each `Filter` call. I decided not to touch it since we will be removing entity sources in #347.

I think as part of #347 we should think about some cache which will be valid for the duration of one resolution.

In the CLI we cache entites into the enttiy source property:

https://github.com/operator-framework/operator-controller/blob/b568602e9daa7df5970090f8f5b791f36f30ddb5/cmd/resolutioncli/entity_source.go#L92-L109

In operator controlle we can not do the same since entity source is resused between reconciliations.

Reply:
> fetching the data from API server on each reconcile & multiple times during one resolution.

Is that true? I thought we had informer caches setup for the catalogd data such that `cl.Get` and `cl.List` would only ever hit the cache (and never the real apiserver)?

Reply:
Ah, you might be right! I totally forgot about informers.